### DISCUSSION TITLE
setupSFC not working at the type system

### DISCUSSION THREAD

**Discussion by angelhdzmultimedia** [vue-macros/vue-macros#667] [Q&A]
Hello

setupSFC is working at runtime, but at the type system it throws errors

![image](https://github.com/vue-macros/vue-macros/assets/10191085/e08b4c97-06f7-48ac-b695-dc02993598eb)

```json
 "dependencies": {
    "@vue-macros/nuxt": "^1.9.30",
    "electron": "^29.3.0",
    "electron-builder": "^24.13.3",
    "nuxt": "3.6.5",
    "nuxt-electron": "^0.7.0",
    "vite-plugin-electron": "^0.28.4",
    "vite-plugin-electron-renderer": "^0.14.5",
    "vue": "^3.4.21",
    "vue-router": "^4.3.0"
  }
```

**Comment by angelhdzmultimedia** [comment]
🤔 Maybe it's the new Vue Language Server extension?

**Reply by sxzz** [reply]
As docs said, 
> ❌ Considering that v2 is not yet stable and some issues remain unresolved, plus adapting to 2.x would mean an almost irreversible upgrade for Vue Macros, we've decided not to support 2.x at this time. In other words, we will continue to support v1.
https://vue-macros.dev/guide/getting-started.html

**Comment by angelhdzmultimedia** [comment]
Reopening. 

Installed the only v1 version available, 1.8.27. Restarted VSCode. Same error. 
![image](https://github.com/vue-macros/vue-macros/assets/10191085/00851f52-cadf-4c21-a4f1-db07e4da5398)

**Comment by angelhdzmultimedia** [comment]
I had a typo. Wrote scriptSetup (i write it a lot) instead of `setupSFC`. Oops!

But here, a repro https://stackblitz.com/edit/github-jw2i4c. Definitely a Vue Official extension issue it seems...

**Comment by angelhdzmultimedia** [comment]
- Installed https://marketplace.visualstudio.com/items?itemName=Vue.volar `1.8.27`
- Closed VSCode
- Opened VSCode
- Same error.
- Checked the examples
- My confi files are all set correctly
- Versions are correct
- Latest Node.js
- Latest VSCode

**Comment by angelhdzmultimedia** [comment]
### Closing

Decided to uninstall Vue Official extension. Issue continues. I'm out of options. Played with nuxt config enabling jsx: preserver or jsx: react-jsx, restarting TS server, disablig TS & JS built-in extension, etc.

Can't get setupSFC to work on the type system. As I mentioned, at runtime it works.

**Comment by angelhdzmultimedia** [comment]
### Solved

No idea what happened. 
First disabled Vue Official extension for the workspace.
Then added `include: ["*.setup.tsx"]` to the macros.setupSFC options in nuxt config.
Then replaced it back to `setupSFC: true`.
Then I noticed the VSCode `pages/index.setup.tsx` file Language Mode was TypeScript, changed it to TypeScript JSX.
Issue got solved.
Enabled Vue Official back.

Funny thing is, in the original project I was working on when opened this Issue, 
I had the language mode set to TypeScript JSX at some point in the debugging process...🤷‍♂️
![image](https://github.com/vue-macros/vue-macros/assets/10191085/ed72cb88-4d0e-4fe7-95a1-3757d61d0b75)

Yay! 🎉💚🔥
![image](https://github.com/vue-macros/vue-macros/assets/10191085/93d3a8bc-ae29-44c1-bbd3-d7c720685643)

**Comment by pythagoreGn** [comment]
bonjour j'ai un probleme avec le build de mon application electron et react mais le probleme viens de canvas quelqu'un pourais ml'aider ?
• rebuilding native dependencies  dependencies=canvas@2.11.2 platform=win32 arch=x64
  ⨯ cannot execute  cause=exit status 1
                    errorOut=npm WARN using --force Recommended protections disabled.
    npm ERR! code 1
    npm ERR! path C:\Users\pythagore\Desktop\boutique projet\reactvierge\node_modules\canvas
    npm ERR! command failed
    npm ERR! command C:\WINDOWS\system32\cmd.exe /d /s /c node-pre-gyp install --fallback-to-build --update-binary
    npm ERR! Les projets sont g�n�r�s individuellement dans cette solution. Pour activer la g�n�ration en parall�le, ajoutez le commutateur "-m".
    npm ERR!   Copying C:/libjpeg-turbo64/bin/jpeg62.dll to C:\Users\pythagore\Desktop\boutique projet\reactvierge\node_modules\canvas\build\Release\
    npm ERR!           1 fichier(s) copi�(s).
    npm ERR!   Backend.cc
    npm ERR! C:\Users\pythagore\Desktop\boutique projet\reactvierge\node_modules\nan\nan.h(688,37): warning C4996: 'v8::Isolate::IdleNotificationDeadline': Use MemoryPressureNotification() to influence the GC schedule. [C:\Users\pythagore\Desktop\boutique projet\reactvierge\node_modules\canvas\build\canvas.vcxproj]    npm ERR! C:\Users\pythagore\Desktop\boutique projet\reactvierge\node_modules\nan\nan.h(2548,6): error C2664: 'void v8::ObjectTemplate::SetAccessor(v8::Local<v8::String>,v8::AccessorGetterCallback,v8::AccessorSetterCallback,v8::Local<v8::Value>,v8::AccessControl,v8::PropertyAttribute,v8::SideEffectType,v8::SideEffectType)'�: impossible de convertir l'argument 2 de 'Nan::imp::NativeGetter' en 'v8::AccessorGetterCallback' [C:\Users\pythagore\Desktop\boutique projet\reactvierge\node_modules\canvas\build\canvas.vcxproj]
    npm ERR! C:\Users\pythagore\Desktop\boutique projet\reactvierge\node_modules\nan\nan.h(2550,7): message : Cette conversion n�cessite un reinterpret_cast, 
un cast de style C ou un cast de style fonction entre parenth�ses. [C:\Users\pythagore\Desktop\boutique projet\reactvierge\node_modules\canvas\build\canvas.vcxproj]
    npm ERR! C:\Users\pythagore\.electron-gyp\28.2.9\include\node\v8-template.h(809,8): message : voir la d�claration de 'v8::ObjectTemplate::SetAccessor' [C:\Users\pythagore\Desktop\boutique projet\reactvierge\node_modules\canvas\build\canvas.vcxproj]
    npm ERR! C:\Users\pythagore\Desktop\boutique projet\reactvierge\node_modules\nan\nan.h(2596,6): error C2664: 'void v8::ObjectTemplate::SetAccessor(v8::Local<v8::String>,v8::AccessorGetterCallback,v8::AccessorSetterCallback,v8::Local<v8::Value>,v8::AccessControl,v8::PropertyAttribute,v8::SideEffectType,v8::SideEffectType)'�: impossible de convertir l'argument 2 de 'Nan::imp::NativeGetter' en 'v8::AccessorGetterCallback' [C:\Users\pythagore\Desktop\boutique projet\reactvierge\node_modules\canvas\build\canvas.vcxproj]
    npm ERR! C:\Users\pythagore\Desktop\boutique projet\reactvierge\node_modules\nan\nan.h(2598,7): message : Cette conversion n�cessite un reinterpret_cast, 
un cast de style C ou un cast de style fonction entre parenth�ses. [C:\Users\pythagore\Desktop\boutique projet\reactvierge\node_modules\canvas\build\canvas.vcxproj]
    npm ERR! C:\Users\pythagore\.electron-gyp\28.2.9\include\node\v8-template.h(809,8): message : voir la d�claration de 'v8::ObjectTemplate::SetAccessor' [C:\Users\pythagore\Desktop\boutique projet\reactvierge\node_modules\canvas\build\canvas.vcxproj]
    npm ERR! Failed to execute 'C:\Program Files\nodejs\node.exe C:\Users\pythagore\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js build --fallback-to-build --update-binary --module=C:\Users\pythagore\Desktop\boutique projet\reactvierge\node_modules\canvas\build\Release\canvas.node --module_name=canvas --module_path=C:\Users\pythagore\Desktop\boutique projet\reactvierge\node_modules\canvas\build\Release --napi_version=9 --node_abi_napi=napi --napi_build_version=0 --node_napi_label=electron-v28.2' (1)
    npm ERR! node-pre-gyp info it worked if it ends with ok
    npm ERR! node-pre-gyp info using node-pre-gyp@1.0.11
    npm ERR! node-pre-gyp info using node@20.12.2 | win32 | x64
    npm ERR! node-pre-gyp http GET https://github.com/Automattic/node-canvas/releases/download/v2.11.2/canvas-v2.11.2-electron-v28.2-win32-unknown-x64.tar.gz 
    npm ERR! node-pre-gyp ERR! install request to https://github.com/Automattic/node-canvas/releases/download/v2.11.2/canvas-v2.11.2-electron-v28.2-win32-unknown-x64.tar.gz failed, reason: getaddrinfo ENOTFOUND github.com
    npm ERR! node-pre-gyp WARN Pre-built binaries not installable for canvas@2.11.2 and electron@28.2.9 (electron-v28.2 ABI, unknown) (falling back to source 
compile with node-gyp)
    npm ERR! node-pre-gyp WARN Hit error request to https://github.com/Automattic/node-canvas/releases/download/v2.11.2/canvas-v2.11.2-electron-v28.2-win32-unknown-x64.tar.gz failed, reason: getaddrinfo ENOTFOUND github.com
    npm ERR! gyp info it worked if it ends with ok
    npm ERR! gyp info using node-gyp@10.1.0
    npm ERR! gyp info using node@20.12.2 | win32 | x64
    npm ERR! gyp info ok
    npm ERR! gyp info it worked if it ends with ok
    npm ERR! gyp info using node-gyp@10.1.0
    npm ERR! gyp info using node@20.12.2 | win32 | x64
    npm ERR! gyp info find Python using Python version 3.9.6 found at "C:\Users\pythagore\AppData\Local\Programs\Python\Python39-32\python.exe"
    npm ERR! gyp info find VS using VS2022 (17.3.32519.111) found at:
    npm ERR! gyp info find VS "C:\Program Files\Microsoft Visual Studio\2022\Preview"
    npm ERR! gyp info find VS run with --verbose for detailed information
    npm ERR! gyp info spawn C:\Users\pythagore\AppData\Local\Programs\Python\Python39-32\python.exe
    npm ERR! gyp info spawn args [
    npm ERR! gyp info spawn args 'C:\\Users\\pythagore\\AppData\\Roaming\\npm\\node_modules\\npm\\node_modules\\node-gyp\\gyp\\gyp_main.py',
    npm ERR! gyp info spawn args 'binding.gyp',
    npm ERR! gyp info spawn args '-f',
    npm ERR! gyp info spawn args 'msvs',
    npm ERR! gyp info spawn args '-I',
    npm ERR! gyp info spawn args 'C:\\Users\\pythagore\\Desktop\\boutique projet\\reactvierge\\node_modules\\canvas\\build\\config.gypi',
    npm ERR! gyp info spawn args '-I',
    npm ERR! gyp info spawn args 'C:\\Users\\pythagore\\AppData\\Roaming\\npm\\node_modules\\npm\\node_modules\\node-gyp\\addon.gypi',
    npm ERR! gyp info spawn args '-I',
    npm ERR! gyp info spawn args 'C:\\Users\\pythagore\\.electron-gyp\\28.2.9\\include\\node\\common.gypi',
    npm ERR! gyp info spawn args '-Dlibrary=shared_library',
    npm ERR! gyp info spawn args '-Dvisibility=default',
    npm ERR! gyp info spawn args '-Dnode_root_dir=C:\\Users\\pythagore\\.electron-gyp\\28.2.9',
    npm ERR! gyp info spawn args '-Dnode_gyp_dir=C:\\Users\\pythagore\\AppData\\Roaming\\npm\\node_modules\\npm\\node_modules\\node-gyp',
    npm ERR! gyp info spawn args '-Dnode_lib_file=C:\\\\Users\\\\pythagore\\\\.electron-gyp\\\\28.2.9\\\\<(target_arch)\\\\node.lib',
    npm ERR! gyp info spawn args '-Dmodule_root_dir=C:\\Users\\pythagore\\Desktop\\boutique projet\\reactvierge\\node_modules\\canvas',
    npm ERR! gyp info spawn args '-Dnode_engine=v8',
    npm ERR! gyp info spawn args '--depth=.',
    npm ERR! gyp info spawn args '--no-parallel',
    npm ERR! gyp info spawn args '--generator-output',
    npm ERR! gyp info spawn args 'C:\\Users\\pythagore\\Desktop\\boutique projet\\reactvierge\\node_modules\\canvas\\build',
    npm ERR! gyp info spawn args '-Goutput_dir=.'
    npm ERR! gyp info spawn args ]
    npm ERR! gyp info ok
    npm ERR! gyp info it worked if it ends with ok
    npm ERR! gyp info using node-gyp@10.1.0
    npm ERR! gyp info using node@20.12.2 | win32 | x64
    npm ERR! gyp info spawn C:\Program Files\Microsoft Visual Studio\2022\Preview\MSBuild\Current\Bin\MSBuild.exe
    npm ERR! gyp info spawn args [
    npm ERR! gyp info spawn args 'build\\binding.sln',
    npm ERR! gyp info spawn args '/clp:Verbosity=minimal',
    npm ERR! gyp info spawn args '/nologo',
    npm ERR! gyp info spawn args '/p:Configuration=Release;Platform=x64'
    npm ERR! gyp info spawn args ]
    npm ERR! gyp ERR! build error
    npm ERR! gyp ERR! stack Error: `C:\Program Files\Microsoft Visual Studio\2022\Preview\MSBuild\Current\Bin\MSBuild.exe` failed with exit code: 1
    npm ERR! gyp ERR! stack at ChildProcess.<anonymous> (C:\Users\pythagore\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\build.js:209:23)   
    npm ERR! gyp ERR! stack at ChildProcess.emit (node:events:518:28)
    npm ERR! gyp ERR! stack at ChildProcess._handle.onexit (node:internal/child_process:294:12)
    npm ERR! gyp ERR! System Windows_NT 10.0.19045
    npm ERR! gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\pythagore\\AppData\\Roaming\\npm\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "build" "--fallback-to-build" "--update-binary" "--module=C:\\Users\\pythagore\\Desktop\\boutique projet\\reactvierge\\node_modules\\canvas\\build\\Release\\canvas.node" "--module_name=canvas" "--module_path=C:\\Users\\pythagore\\Desktop\\boutique projet\\reactvierge\\node_modules\\canvas\\build\\Release" "--napi_version=9" "--node_abi_napi=napi" "--napi_build_version=0" "--node_napi_label=electron-v28.2"
    npm ERR! gyp ERR! cwd C:\Users\pythagore\Desktop\boutique projet\reactvierge\node_modules\canvas
    npm ERR! gyp ERR! node -v v20.12.2
    npm ERR! gyp ERR! node-gyp -v v10.1.0
    npm ERR! gyp ERR! not ok
    npm ERR! node-pre-gyp ERR! build error
    npm ERR! node-pre-gyp ERR! stack Error: Failed to execute 'C:\Program Files\nodejs\node.exe C:\Users\pythagore\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js build --fallback-to-build --update-binary --module=C:\Users\pythagore\Desktop\boutique projet\reactvierge\node_modules\canvas\build\Release\canvas.node --module_name=canvas --module_path=C:\Users\pythagore\Desktop\boutique projet\reactvierge\node_modules\canvas\build\Release --napi_version=9 --node_abi_napi=napi --napi_build_version=0 --node_napi_label=electron-v28.2' (1)
    npm ERR! node-pre-gyp ERR! stack     at ChildProcess.<anonymous> (C:\Users\pythagore\Desktop\boutique projet\reactvierge\node_modules\@mapbox\node-pre-gyp\lib\util\compile.js:89:23)
    npm ERR! node-pre-gyp ERR! stack     at ChildProcess.emit (node:events:518:28)
    npm ERR! node-pre-gyp ERR! stack     at maybeClose (node:internal/child_process:1105:16)
    npm ERR! node-pre-gyp ERR! stack     at ChildProcess._handle.onexit (node:internal/child_process:305:5)
    npm ERR! node-pre-gyp ERR! System Windows_NT 10.0.19045
    npm ERR! node-pre-gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\pythagore\\Desktop\\boutique projet\\reactvierge\\node_modules\\@mapbox\\node-pre-gyp\\bin\\node-pre-gyp" "install" "--fallback-to-build" "--update-binary"
    npm ERR! node-pre-gyp ERR! cwd C:\Users\pythagore\Desktop\boutique projet\reactvierge\node_modules\canvas
    npm ERR! node-pre-gyp ERR! node -v v20.12.2
    npm ERR! node-pre-gyp ERR! node-pre-gyp -v v1.0.11
    npm ERR! node-pre-gyp ERR! not ok

    npm ERR! A complete log of this run can be found in: C:\Users\pythagore\AppData\Local\npm-cache\_logs\2024-04-27T03_55_57_477Z-debug-0.log

                    command='C:\Program Files\nodejs\node.exe' 'C:\Users\pythagore\AppData\Roaming\npm\node_modules\npm\bin\npm-cli.js' rebuild canvas@2.11.2 
                    workingDir=

**Comment by angelhdzmultimedia** [comment]
@pythagoreGn Je suis curieux. Votre problème n'est même pas lié à la bibliothèque vue-macros. Même pas lié à Vue ! Comment êtes-vous arrivé ici ?

**Reply by pythagoreGn** [reply]
@angelhdzmultimedia peut tu m'aider ?

**Comment by pythagoreGn** [comment]
désolé frère je cherche juste une solution à mon problème si tu as une
solution aide moi stp

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Sans
virus.www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

**Comment by angelhdzmultimedia** [comment]
> désolé frère je cherche juste une solution à mon problème si tu as une solution aide moi stp

Sure. I'm always willing to help. I was just curious at the fact that you seeked help in a very non-related Issue. 👀🤔

**Comment by angelhdzmultimedia** [comment]
@pythagoreGn I would first start by deleting node_modules, package-lock.json (or the lock file for your package manager), then `npm install`, then update dependencies to their latest versions (npm update). If that doesn't fix it, check the canvas package, downgrade many versions down. If the problem persist with at least 2 major versions down, then is not canvas package's fault, then diagnose electron or electron-builder. Also check that you are using recommended Node version.

It's hard to diagnose without enough context.
